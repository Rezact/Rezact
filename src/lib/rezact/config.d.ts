import { UserConfig } from "vite";

interface defOptions {
  routes?: any[];
  options?: {
    useMdx?: boolean;
  };
}

export declare const configRezact: (item: defOptions) => UserConfig;
