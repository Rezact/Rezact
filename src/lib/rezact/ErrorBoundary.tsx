import {
  // addPostCreateComponentHook,
  // addPreCreateComponentHook,
  errors,
  render,
} from ".";
import { Signal } from "./signals";

// addPreCreateComponentHook((tagName, attributes, hookContext) => {
//   if (tagName !== ErrorBoundary) return;
//   // console.log({ tagName: tagName.name, attributes, hookContext });
// });

// addPostCreateComponentHook((newComp, tagName, attributes, hookContext) => {
//   if (tagName !== ErrorBoundary) return;
//   // console.log({ newComp, tagName, attributes, hookContext });
// });

function DefaultError(props: any) {
  return (
    <div>
      <p>Error</p>
      <textarea>{props.errors.map((err) => err.stack)}</textarea>
    </div>
  );
}

export function ErrorBoundary(props: any) {
  props.fallback = props.fallback || DefaultError;

  const fb =
    typeof props.fallback === "function"
      ? props.fallback
      : () => props.fallback;

  const renderedTree = new Signal(<span />);

  if (errors.length > 0) {
    render(renderedTree, fb, { errors });
    errors.length = 0;
  } else {
    render(renderedTree, () => props.children);
  }

  return renderedTree;
}
