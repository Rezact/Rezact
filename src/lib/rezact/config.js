import { rezact } from "./vite-plugin.js";
import { rezact_mdx } from "./vite-mdx-plugin.js";
import remarkFrontmatter from "remark-frontmatter";
import remarkMdxFrontmatter from "remark-mdx-frontmatter";
import rehypeHighlight from "rehype-highlight";
import mdx from "@mdx-js/rollup";
import { addRoutesToInput, rezactBuild } from "./vite-build-plugin.js";
import { defineConfig } from "vite";

const config = defineConfig({
  resolve: {
    alias: {
      src: "/src",
      rezact: "@rezact/rezact",
    },
  },
  build: {
    target: "esnext",
    modulePreload: {
      polyfill: false,
    },
    rollupOptions: {},
  },
  esbuild: {
    jsxFactory: "xCreateElement",
    jsxFragment: "xFragment",
  },
  plugins: [],
});

export function configRezact({ routes, options }) {
  const useMdx = options?.useMdx ?? true;
  if (!config.plugins) config.plugins = [];

  if (useMdx)
    config.plugins.push(
      mdx({
        pragma: "r.xCreateElement",
        pragmaFrag: "r.xFragment",
        jsxRuntime: "classic",
        pragmaImportSource: "@rezact/rezact/mdx",
        remarkPlugins: [
          remarkFrontmatter,
          [remarkMdxFrontmatter, { name: "fm" }],
        ],
        rehypePlugins: [rehypeHighlight],
      })
    );

  config.plugins.push(rezact());

  if (useMdx) config.plugins.push(rezact_mdx());

  if (routes) {
    addRoutesToInput(routes, config);
    config.plugins.push(rezactBuild({ routes }));
  }

  return config;
}
