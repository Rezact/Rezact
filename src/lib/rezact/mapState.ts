import { Signal, afterSignalRenderHook, effect } from "./signals";
import { inRender, createComponent, isArray } from ".";

function isIgnoredInstance(obj: any): boolean {
  return (
    obj instanceof Window ||
    obj instanceof Document ||
    obj instanceof HTMLElement ||
    obj instanceof Node
  );
}

const skipKeys = new Set([
  "elmRef",
  "idxState",
  "deps",
  "subs",
  "func",
  "associatedState",
  "state",
  "computed",
]);

function findNestedStates(obj, results = []) {
  if (isIgnoredInstance(obj)) return [];

  // Check if the input is an object
  if (typeof obj === "object" && obj !== null) {
    // If it's an array, iterate over each element
    if (Array.isArray(obj)) {
      for (let item of obj) {
        findNestedStates(item, results);
      }
    } else {
      // If it's an object, iterate over each property
      for (let key in obj) {
        if (key.startsWith("$")) {
          results.push(obj[key]);
        }
        if (skipKeys.has(key)) continue;
        if (isIgnoredInstance(obj[key])) continue;

        findNestedStates(obj[key], results);
      }
    }
  }
  return results;
}

function subscribeToNestedStates(item, mapStateObj) {
  const nestedStates = findNestedStates(item);
  nestedStates.forEach((state) => {
    item.nestedSubscribed = true;

    state.subscribe(
      () => {
        if (
          mapStateObj.deps &&
          mapStateObj.deps.length > 0 &&
          mapStateObj.deps[0] instanceof MapSignal
        ) {
          mapStateObj.deps[0].refresh();
        } else {
          mapStateObj.refresh();
        }
      },
      { elm: item.elmRef }
    );
  });
}

export class MapSignal<T> extends Signal<T> {
  elmRefCache: any = new Map();
  refreshTimer: any;
  clearCacheTimer: any;
  push: (...items: T[]) => {};
  map: (
    callbackfn: (value: T, index: number, array: T[]) => unknown,
    thisArg?: any
  ) => {};

  constructor(st: T[]) {
    super(st as T);
    ["push", "pop", "splice", "shift", "unshift", "reduce", "map"].forEach(
      (item) => {
        this[item] = (...args) => {
          this.value[item](...args);
          this.makeChildrenSignals();
          this.updateList(this.func);
          this.alertSubs(this.value);
          if (!inRender) afterSignalRenderHook();
        };
      }
    );
    this.makeChildrenSignals();
  }

  removeStaleElmRefCacheItems() {
    if (this.clearCacheTimer) clearTimeout(this.clearCacheTimer);
    this.clearCacheTimer = setTimeout(() => {
      for (let [key, value] of this.elmRefCache) {
        if (!value.isConnected) {
          this.elmRefCache.delete(key);
        }
      }
      this.deps &&
        this.deps.forEach((dep) => {
          if (dep instanceof MapSignal) {
            dep.removeStaleElmRefCacheItems();
          }
        });
    }, 10);
  }

  Map(func: any) {
    if (!func.args) {
      this.func = func;
      const _args = func.toString().split("=>")[0].replaceAll(" ", "");
      const args = _args.slice(1, _args.length - 1).split(",");
      func.args = args;
    }
    this.updateList(func);
    return this.value;
  }

  updateList(func) {
    if (func === undefined) return;
    const len = this.value.length;
    for (let idx = 0; idx < len; idx++) {
      let item = this.value[idx];
      if (!item.state) this.value[idx] = item = new Signal(item);
      item.__private_idx = idx;
      if (func.args[1]) {
        if (!item.idxState) {
          item.idxState = new Signal(0);
        }
        item.idxState.set(idx);
      }

      let cachedElmRef = this.elmRefCache.get(item);
      if (cachedElmRef?.isConnected === false) {
        this.elmRefCache.delete(item);
        cachedElmRef = undefined;
      }

      if (
        item.elmRef?.isConnected === false ||
        item.elmRef?.isConnected === undefined
      )
        item.elmRef = undefined;

      item.elmRef = cachedElmRef || createComponent({ func, item });
      if (!cachedElmRef) this.elmRefCache.set(item, item.elmRef);

      if (!item.nestedSubscribed) {
        subscribeToNestedStates(item, this);
      }

      isArray(item.elmRef)
        ? item.elmRef.forEach((elm) => (elm.associatedState = item))
        : (item.elmRef.associatedState = item);
    }
  }

  makeChildrenSignals() {
    if (!this.value) return;
    const len = this.value.length;
    for (let idx = 0; idx < len; idx++) {
      let item = this.value[idx];
      if (!item.state) this.value[idx] = item = new Signal(item);
      item.__private_idx = idx;

      if (!item.nestedSubscribed) {
        subscribeToNestedStates(item, this);
      }
    }
  }

  deleteValue(valToDelete: any) {
    const index = valToDelete.__private_idx || this.value.indexOf(valToDelete);
    if (index < 0) return;
    if (valToDelete.elmRef && valToDelete.elmRef instanceof HTMLElement) {
      this.elmRefCache.delete(valToDelete);
      valToDelete.elmRef.remove();
    }
    this.value.splice(index, 1);
    this.alertSubs(this.value);
    setTimeout(() => this.updateList(this.func), 0);
  }

  toJSON() {
    return this.value.map((thisVal) => {
      if (typeof thisVal === "object" && !thisVal.state && !thisVal.get)
        return thisVal;
      if (thisVal.toJSON) return thisVal.toJSON();
      thisVal.get ? thisVal.get() : thisVal;
    });
  }

  refresh = () => {
    if (this.refreshTimer) clearTimeout(this.refreshTimer);
    this.refreshTimer = setTimeout(() => {
      this.updateList(this.func);
      this.alertSubs(this.value);
    }, 10);
  };
}

export const mapEffect = (func, deps) => {
  const computed: any = effect(func, deps);
  computed.deps = deps;
  computed.mapStateObj = false;
  return computed;
};
