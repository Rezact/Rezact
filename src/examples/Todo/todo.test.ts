import { describe, expect, it } from "vitest";
import { screen, waitFor } from "@testing-library/dom";
import { render } from "rezact";
import { Page } from "./Todo";
import userEvent from "@testing-library/user-event";
import { delay } from "src/lib/utils";

const user = userEvent.setup();

const testDiv = document.createElement("div");
document.body.appendChild(testDiv);

describe("Todo suite", () => {
  it("Renders Todo List", async () => {
    render(testDiv, Page);
    await delay(100);
    expect(screen.getByText("Todo List")).not.toBeNull();
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Filters Todo List", async () => {
    const todos = document.querySelectorAll("div[style]");
    expect(todos.length).toBe(3);

    await user.click(screen.getByText("Show Completed"));
    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(1);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    await user.click(screen.getByText("Show All"));
    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(3);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    await user.click(screen.getByText("Show Todo"));
    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(2);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Adds Todo", async () => {
    const newTodoInput = screen.getByPlaceholderText("Add Todo");
    const addButton = screen.getByRole("button", { name: "Add" });

    await user.type(newTodoInput, "New Todo");
    await user.click(addButton);

    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(3);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    await user.type(newTodoInput, "Another Todo");
    await user.click(addButton);

    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(4);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    await user.type(newTodoInput, "Doing it again");
    await user.click(addButton);

    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(5);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Completes Todo", async () => {
    let todos = document.querySelectorAll("div[style]");
    expect(todos.length).toBe(5);

    const completeButton = todos[0].querySelector("input[type=checkbox]");
    await user.click(completeButton);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(4);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    todos = document.querySelectorAll("div[style]");
    const completeButton2 = todos[2].querySelector("input[type=checkbox]");
    await user.click(completeButton2);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(3);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    todos = document.querySelectorAll("div[style]");
    const completeButton3 = todos[2].querySelector("input[type=checkbox]");
    await user.click(completeButton3);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(2);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Edits Todo", async () => {
    let todos = document.querySelectorAll("div[style]");
    expect(todos.length).toBe(2);
    let span = todos[0].querySelectorAll("span")[1];
    expect(span.innerText).toBe("1 - Build something awesome");

    let inputs = todos[0].querySelectorAll("input");
    await user.dblClick(span);
    await waitFor(() => {
      inputs = todos[0].querySelectorAll("input");
      expect(inputs.length).toBe(2);
    });

    await user.clear(inputs[0]);
    await user.type(inputs[0], "Edited");
    await user.keyboard("{Enter}");
    await waitFor(() => {
      span = todos[0].querySelectorAll("span")[1];
      expect(span.innerText).toBe("1 - Edited");
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Deletes Todo", async () => {
    let todos = document.querySelectorAll("div[style]");
    expect(todos.length).toBe(2);

    const deleteButton = todos[0].querySelector("button");
    await user.click(deleteButton);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(1);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    todos = document.querySelectorAll("div[style]");
    const deleteButton2 = todos[0].querySelector("button");
    await user.click(deleteButton2);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(0);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Completed Todos has more items", async () => {
    await user.click(screen.getByText("Show Completed"));
    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(4);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Unchecks Completed Todo", async () => {
    let todos = document.querySelectorAll("div[style]");
    expect(todos.length).toBe(4);

    const completeButton = todos[0].querySelector("input[type=checkbox]");
    await user.click(completeButton);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(3);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    todos = document.querySelectorAll("div[style]");
    const completeButton2 = todos[0].querySelector("input[type=checkbox]");
    await user.click(completeButton2);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(2);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Show All has more items", async () => {
    await user.click(screen.getByText("Show All"));
    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(4);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Clears Completed Todos", async () => {
    await user.click(screen.getByText("Remove Completed"));
    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      expect(todos.length).toBe(2);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Checks All Todos", async () => {
    await user.click(screen.getByText("Check All"));
    await waitFor(() => {
      const todos = document.querySelectorAll("div[style]");
      const chk1 = todos[0].querySelector(
        "input[type=checkbox]"
      ) as HTMLInputElement;
      const chk2 = todos[1].querySelector(
        "input[type=checkbox]"
      ) as HTMLInputElement;
      expect(chk1.checked).toBe(true);
      expect(chk2.checked).toBe(true);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });

  it("Deletes Todo Items", async () => {
    let todos = document.querySelectorAll("div[style]");
    expect(todos.length).toBe(2);

    const deleteButton = todos[0].querySelector("button");
    await user.click(deleteButton);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(1);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();

    todos = document.querySelectorAll("div[style]");
    const deleteButton2 = todos[0].querySelector("button");
    await user.click(deleteButton2);
    await waitFor(() => {
      const test = document.querySelectorAll("div[style]");
      expect(test.length).toBe(0);
    });
    await delay(100);
    expect(testDiv.innerHTML).toMatchSnapshot();
  });
});
