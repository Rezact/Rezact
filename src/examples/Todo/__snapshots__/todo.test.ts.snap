// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todo suite > Adds Todo 1`] = `"<h1>Todo List</h1><p>4 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - Build something awesome</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>3 - New Todo</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Adds Todo 2`] = `"<h1>Todo List</h1><p>5 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - Build something awesome</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>3 - New Todo</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>4 - Another Todo</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Adds Todo 3`] = `"<h1>Todo List</h1><p>6 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - Build something awesome</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>3 - New Todo</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>4 - Another Todo</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>5 - Doing it again</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Checks All Todos 1`] = `"<h1>Todo List</h1><p>2 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style="text-decoration: line-through;"><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>2 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Clears Completed Todos 1`] = `"<h1>Todo List</h1><p>2 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Completed Todos has more items 1`] = `"<h1>Todo List</h1><p>4 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style="text-decoration: line-through;"><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>2 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>3 - Another Todo</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>4 - Doing it again</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Completes Todo 1`] = `"<h1>Todo List</h1><p>6 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Build something awesome</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - New Todo</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>3 - Another Todo</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>4 - Doing it again</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Completes Todo 2`] = `"<h1>Todo List</h1><p>6 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Build something awesome</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - New Todo</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>3 - Doing it again</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Completes Todo 3`] = `"<h1>Todo List</h1><p>6 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Build something awesome</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - New Todo</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Deletes Todo 1`] = `"<h1>Todo List</h1><p>5 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - New Todo</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Deletes Todo 2`] = `"<h1>Todo List</h1><p>4 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Deletes Todo Items 1`] = `"<h1>Todo List</h1><p>1 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style="text-decoration: line-through;"><span></span><span>1 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Deletes Todo Items 2`] = `"<h1>Todo List</h1><p>0 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Edits Todo 1`] = `"<h1>Todo List</h1><p>6 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Edited</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - New Todo</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Filters Todo List 1`] = `"<h1>Todo List</h1><p>3 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style="text-decoration: line-through;"><span></span><span>1 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Filters Todo List 2`] = `"<h1>Todo List</h1><p>3 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>2 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>3 - Build something awesome</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Filters Todo List 3`] = `"<h1>Todo List</h1><p>3 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - Build something awesome</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Renders Todo List 1`] = `"<h1>Todo List</h1><p>3 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>2 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>3 - Build something awesome</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Show All has more items 1`] = `"<h1>Todo List</h1><p>4 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style=""><span></span><span>1 - Learn Rezact</span><input type="checkbox"><button>X</button></div><div style=""><span></span><span>2 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>3 - Another Todo</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>4 - Doing it again</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Unchecks Completed Todo 1`] = `"<h1>Todo List</h1><p>4 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style="text-decoration: line-through;"><span></span><span>1 - Learn TypeScript</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>2 - Another Todo</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>3 - Doing it again</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;

exports[`Todo suite > Unchecks Completed Todo 2`] = `"<h1>Todo List</h1><p>4 Todos</p><fieldset><legend>Filter:</legend><div><input type="radio" id="all" value="all"><label for="all">Show All</label></div><div><input type="radio" id="completed" value="completed"><label for="completed">Show Completed</label></div><div><input type="radio" id="todo" value="todo"><label for="todo">Show Todo</label></div></fieldset><!--start map--><div style="text-decoration: line-through;"><span></span><span>1 - Another Todo</span><input type="checkbox"><button>X</button></div><div style="text-decoration: line-through;"><span></span><span>2 - Doing it again</span><input type="checkbox"><button>X</button></div><!--end map--><form><input placeholder="Add Todo"><button>Add</button></form><button>Remove Completed</button><button>Check All</button><!--signal-->"`;
